devito/operator.py

```diff
@@ -154,14 +150,23 @@ def __init__(self, nt, shape, dtype=np.float32, stencils=[],
         self.shape = shape
         self.spc_border = spc_border
         self.symbol_to_data = {}
-        for param in self.input_params + self.output_params:
+        for param in self.signature:
             self.propagator.add_devito_param(param)
             self.symbol_to_data[param.name] = param
         self.propagator.stencils = self.stencils
         self.propagator.factorized = factorized
         for name in factorized.keys():
             self.propagator.factorized[name] = expr_indexify(factorized[name])
 
+    @property
+    def signature(self):
+        """List of data object parameters that define the operator signature
+
+        :returns: List of unique input and output data objects
+        """
+        return self.input_params + [param for param in self.output_params
```

Comment:
The implicit repetition comes from auto-detecting the params in stencils that update a field variable, ie. `u[t] = u[t-1]`, where the same variable appears in input and output params.